README.md
# FOSSology

[![Travis-CI Build Status](https://travis-ci.org/fossology/fossology.svg?branch=master)](https://travis-ci.org/fossology/fossology)
[![Stories in Ready](https://badge.waffle.io/fossology/fossology.svg?label=ready&title=Ready)](http://waffle.io/fossology/fossology)
[![Coverage Status](https://coveralls.io/repos/github/fossology/fossology/badge.svg?branch=master)](https://coveralls.io/github/fossology/fossology?branch=master)

## About

FOSSology is a open source license compliance software system and toolkit.  As a toolkit you can run license, copyright and export control scans from the command line.  As a system, a database and web ui are provided to give you a compliance workflow. In one click you can generate an SPDX file, or a ReadMe with all the copyrights notices from your software. FOSSology deduplication means that you can scan an entire distro, rescan a new version, and only the changed files will get rescanned. This is a big time saver for large projects.

[Check out Who Uses FOSSology!](http://www.fossology.org/projects/fossology/wiki/WhoUsesFOSSology)

FOSSology does not give legal advice.
http://fossology.org/

## Requirements

The PHP versions 5.5.9 to 5.6.x are supported to work for FOSSology. FOSSology requires Postgresql as database server and apache httpd 2.4 as web server. These and more dependencies are installed by `utils/fo-installdeps`.

## Installation

FOSSology should work with many Linux distributions.

See https://github.com/fossology/fossology/releases for source code download of the releases.

For installation instructions see [Github Wiki](https://github.com/fossology/fossology/wiki)

## Docker

FOSSology comes with a Dockerfile allowing the containerized execution
both as single instance or in combination with an external PostgreSQL database.
**Note:** It is strongly recommended to use an external database for production
use, since the the standalone image does not take care of data persistency.

A pre-built Docker image is available from [Docker Hub](https://hub.docker.com/r/fossology/fossology/) and can be run using following command:
``` sh
docker run -p 8081:80 fossology/fossology
```

The docker image can then be used using http://IP_OF_DOCKER_HOST:8081/repo user fossy passwd fossy.

Execution with external database container can be done using Docker Compose.
The Docker Compose file is located under the `/install` folder can can be run using following command:
``` sh
cd install
docker-compose up
```

The Docker image allows configuration of it's database connection over a set of environment variables.

- **FOSSOLOGY_DB_HOST:** Hostname of the PostgreSQL database server.
  An integrated PostgreSQL instance is used if not defined or set to `localhost`.
- **FOSSOLOGY_DB_NAME:** Name of the PostgreSQL database. Defaults to `fossology`.
- **FOSSOLOGY_DB_USER:** User to be used for PostgreSQL connection. Defaults to `fossy`.
- **FOSSOLOGY_DB_PASSWORD:** Password to be used for PostgreSQL connection. Defaults to `fossy`.

## Vagrant

FOSSology comes with a VagrantFile that can be used to create an isolated environment for FOSSology and its dependencies.

**Pre-requisites:**  Vagrant >= 2.x and Virtualbox >= 5.2.x

**Steps:**

```    
git clone https://github.com/fossology/fossology
cd fossology/
vagrant up
```

The server must be ready at [http://localhost:8081/repo/](http://localhost:8081/repo/) and user can login the credentials using 

```
user: fossy
pass: fossy
```

## Documentation

We are currently migrating our documentation to github.  At this stage you can find general documentation at:
http://www.fossology.org/projects/fossology/wiki/User_Documentation
and developer docs here on [github](https://github.com/fossology/fossology/wiki)

## Support

Mailing lists, FAQs, Release Notes, and other useful info is available
by clicking the documentation tab on the project website. We encourage
all users to join the mailing list and participate in discussions.
There is also a #fossology IRC channel on the freenode IRC network if
you'd like to talk to other FOSSology users and developers.
See [Contact Us](https://www.fossology.org/get-started)

## Contributing

We really like contributions in several forms, see [CONTRIBUTING.md](CONTRIBUTING.md)

## Licensing

The original FOSSology source code and associated documentation
including these web pages are Copyright (C) 2007-2012 HP Development
Company, L.P. In the past years, other contributors added source code 
and documentation to the project, see the NOTICES file or the referring
files for more information.

Any modifications or additions to source code or documentation
contributed to the FOSSology project are Copyright (C) the contributor,
and should be noted as such in the comments section of the modified file(s).

FOSSology is licensed under [GPL-2.0](https://tldrlegal.com/license/gnu-general-public-license-v2)

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; version 2 of the License

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

Exception:

All of the FOSSology source code is licensed under the terms of the GNU
General Public License version 2, with the following exceptions:

libfossdb and libfossrepo libraries are licensed under the terms of
the GNU Lesser General Public License version 2.1, [LGPL-2.1](https://tldrlegal.com/license/gnu-lesser-general-public-license-v2.1-(lgpl-2.1)).

    This library are free software; you can redistribute it and/or
    modify it under the terms of the GNU Lesser General Public
    License as published by the Free Software Foundation; either
    version 2.1 of the License.

    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public
    License along with this library; if not, write to the Free Software
    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
    02110-1301  USA

Please see the files COPYING and COPYING.LGPL included with this
software for the full text of these licenses.

## Screenshots

![Browsing](examples/Browseb.jpg)

![LicenseBrowser](examples/LicenseBrowserb.jpg)

![Concluding a license](examples/Concludeb.jpg)



debian/changelog
fossology (3.3.0-1) unstable; urgency=low

  * New upstream release
  * https://github.com/fossology/fossology/releases/tag/3.3.0 for details

 -- Michael C. Jaeger <michael.c.jaeger@siemens.com>  Wed, 17 Apr 2018 13:57:55 +0800

fossology (3.2.0-1) unstable; urgency=low

  * New upstream release
  * https://github.com/fossology/fossology/releases/tag/3.2.0 for details

 -- Michael C. Jaeger <michael.c.jaeger@siemens.com>  Thu, 8 Feb 2018 15:35:55 +0800

fossology (3.2.0~rc1-1) unstable; urgency=low

  * New upstream release
  * https://github.com/fossology/fossology/releases/tag/3.2.0rc1 for details

 -- Michael C. Jaeger <michael.c.jaeger@siemens.com>  Sat, 20 Oct 2017 15:35:55 +0800

fossology (3.1.0-1) unstable; urgency=low

  * New upstream release
  * see https://github.com/fossology/fossology/releases/tag/3.1.0 for details

 -- Michael C. Jaeger <michael.c.jaeger@siemens.com>  Sat, 21 Apr 2017 15:35:55 +0800

fossology (3.1.0~rc2-1) unstable; urgency=low

  * New upstream release
  * see https://github.com/fossology/fossology/releases/tag/3.1.0rc2 for details

 -- maximilian.huber <maximilian.huber@tngtech.com>  Sat, 21 May 2016 15:35:55 +0800

fossology (3.1.0~rc1-1) unstable; urgency=low

  * New upstream release
  * see https://github.com/fossology/fossology/releases/tag/3.1.0rc1 for details

 -- maximilian.huber <maximilian.huber@tngtech.com>  Fri, 15 Apr 2016 15:35:55 +0800

fossology (3.0.0-1) unstable; urgency=low

  * New upstream release
  * see https://github.com/fossology/fossology/releases/tag/3.0.0 for details

 -- maximilian.huber <maximilian.huber@tngtech.com>  Thu, 05 Nov 2015 15:35:55 +0800

fossology (2.6.2-1) unstable; urgency=low

  * New upstream release

 -- larainema <larainema@localhost.localdomain>  Mon, 12 Jan 2015 15:35:55 +0800

fossology (2.6.1-1) unstable; urgency=low

  * New upstream release

 -- larainema <larainema@localhost.localdomain>  Wed, 01 Oct 2014 15:35:55 +0800

fossology (2.6.0-1) unstable; urgency=low

  * New upstream release

 -- larainema <larainema@localhost.localdomain>  Fri, 19 Sep 2014 15:35:55 +0800

fossology (2.5.0-1) unstable; urgency=low

  * New upstream release

 -- larainema <larainema@localhost.localdomain>  Tue, 08 Apr 2014 15:35:55 +0800
 
fossology (2.4.0-1) unstable; urgency=low

  * New upstream release

 -- larainema <larainema@localhost.localdomain>  Tue, 14 Jan 2014 15:35:55 +0800

fossology (2.3.0-1) unstable; urgency=low

  * New upstream release

 -- larainema <larainema@localhost.localdomain>  Web, 25 Sep 2013 15:35:55 +0800

fossology (2.2.0-1) unstable; urgency=low

  * New upstream release

 -- larainema <larainema@localhost.localdomain>  Fri, 28 Jun 2013 15:35:55 +0800
 
fossology (2.1.0-1) unstable; urgency=low

  * New upstream release

 -- madong <madong@bl480c1>  Thu, 01 Nov 2012 15:35:55 +0800

fossology (2.0.0-1) unstable; urgency=low

  * New upstream release
  * New package structure

 -- madong <madong@bl480c1>  Fri, 08 Jun 2012 15:35:55 +0800

fossology (2.0.0~rc3-1) unstable; urgency=low

  * Release candidate 3 for v2.0

 -- madong <madong@bl480c-1>  Fri, 01 Jun 2012 15:22:47 +0800

fossology (2.0.0~rc1-1) unstable; urgency=low

  * The first release candidate for v2.0

 -- madong <madong@bl480c-1>  Wed, 02 May 2012 15:22:47 +0800

fossology (1.4.1-1) unstable; urgency=low

  * New upstream release
  * Bug Fix release
  * Use project-builder to create deb

 -- Mark Donohoe <markd@localhost.localdomain>  Thu, 30 Jun 2011 17:27:23 -0700

fossology (1.4.0-1) unstable; urgency=low

  * New upstream release
  * New Comparison tool
  * Simplified User Interface
  * authentication using Siteminder
  * Multiple file upload from URL
  * Bugs fix
  
 -- Mark Donohoe <markd@localhost.localdomain>  Thu, 05 May 2011 17:55:09 -0700

fossology (1.3.0-1) unstable; urgency=low

  * New upstream release
  * Groups
  * File Tagging
  * Copyright agent replaced
  * Fixed a cp2foss authentication bug
  * fixed unpack defects and made some improvements

 -- Mark Donohoe <markd@steezy>  Fri, 07 Jan 2011 10:35:08 -0800

fossology (1.2.1~SVNBUILD-1) unstable; urgency=low

  * unofficial build version

 -- Matt Taggart <taggart@debian.org>  Fri, 06 Aug 2010 16:53:49 -0700

fossology (1.2.0-2) unstable; urgency=low

  * added upstream fixes to 1.2.0 that will go in 1.2.1
    - fix pkgagent bug, svn#3363
    - free temp name string on fopen error condition, svn#3362
    - fix too many open files bug, svn#3361
    - add error checking to fopen, svn#3359
    - fix pkgagent dsc parsing, svn#3358
    - fix copyright wrong variable name, svn#3355
    - bucket agent fixes, svn#3351-3353
    - fix missing copyright database, svn#3348
    - init script restart output, svn#3346

 -- Matt Taggart <taggart@debian.org>  Fri, 06 Aug 2010 15:12:18 -0700

fossology (1.2.0-1) unstable; urgency=low

  * New upstream release
  * Added build-dep on python-pyrex, python, and python-dev, and
     fossology-common depends on python due to libfosspython.
  * Add python Depends to agent packages.
  * Update standards version.
  * Make references to licenses in copyright more specific.
  * Add upstream SVN#3338 fix for ununpack p7zip.

 -- Matt Taggart <taggart@debian.org>  Tue, 20 Jul 2010 16:20:31 -0700

fossology (1.1.0-2) unstable; urgency=low

  * Acknowledge NMU
   - Replace depends on xpdf-utils with poppler-utils (Closes: #550653 )
  * fix description to actually be a proper sentence (Closes: #539716 )
  * change fossology-db to better explain that it will create the db
    (Closes: #540252 )
  * fix typo in init script Provides (will be in next upstream)
      (Closes: #541854 )
  * Build against PostgreSQL 8.4. (Closes: #559582 )
  * patch from upstream SVN 2848 to build libs with -fPIC
  * patch from upstream SVN 2484 to make selftest not check agent timestamps
  * fix some lintian warnings and update to Policy 3.8.4
  * patch from upstream SVN 2585 and 2605 to fix fo_notify


src/www/ui/user-add.php
<?php
/***********************************************************
 Copyright (C) 2008-2013 Hewlett-Packard Development Company, L.P.

 This program is free software; you can redistribute it and/or
 modify it under the terms of the GNU General Public License
 version 2 as published by the Free Software Foundation.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License along
 with this program; if not, write to the Free Software Foundation, Inc.,
 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 ***********************************************************/

use Fossology\Lib\Db\DbManager;

define("TITLE_user_add", _("Add A User"));

class user_add extends FO_Plugin {

  /** @var DbManager */
  private $dbManager;

  function __construct()
  {
    $this->Name = "user_add";
    $this->Title = TITLE_user_add;
    $this->MenuList = "Admin::Users::Add";
    $this->DBaccess = PLUGIN_DB_ADMIN;
    parent::__construct();
    $this->dbManager = $GLOBALS['container']->get('db.manager');
  }

  /**
   * \brief Add a user.
   * 
   * \return NULL on success, string on failure.
   */
  function Add() {
     
    global $PG_CONN;

    if (!$PG_CONN) {
      DBconnect();
      if (!$PG_CONN)
      {
        $text = _("NO DB connection!");
        echo "<pre>$text\n</pre>";
      }
    }

    /* Get the parameters */
    $User = str_replace("'", "''", GetParm('username', PARM_TEXT));
    $User = trim($User);
    $Pass = GetParm('pass1', PARM_TEXT);
    $Pass2 = GetParm('pass2', PARM_TEXT);
    $Seed = rand() . rand();
    $Hash = sha1($Seed . $Pass);
    $Desc = str_replace("'", "''", GetParm('description', PARM_TEXT));
    $Perm = GetParm('permission', PARM_INTEGER);
    $Folder = GetParm('folder', PARM_INTEGER);
    $Email_notify = GetParm('enote', PARM_TEXT);
    $Email = str_replace("'", "''", GetParm('email', PARM_TEXT));
    $agentList = userAgents();
    $default_bucketpool_fk = GetParm('default_bucketpool_fk', PARM_INTEGER);


    /* Make sure username looks valid */
    if (empty($User)) {
      $text = _("Username must be specified. Not added.");
      return ($text);
    }
    /* limit the user name size to 64 characters when creating an account */
    if (strlen($User) > 64)
    {
      $text = _("Username exceed 64 characters. Not added.");
      return ($text);
    }
    /* Make sure password matches */
    if ($Pass != $Pass2) {
      $text = _("Passwords did not match. Not added.");
      return ($text);
    }
    
    if(empty($Email))
    {
      $text = _("Email must be specified. Not added.");
      return ($text);
    }

    /* Make sure email looks valid */
    if (!filter_var($Email, FILTER_VALIDATE_EMAIL))
    {
      $text = _("Invalid email address.  Not added.");
      return ($text);
    }
    
    /* Make sure email is unique */
    $email_count = $this->dbManager->getSingleRow("SELECT COUNT(*) as count FROM users WHERE user_email = $1 LIMIT 1;", array($Email))["count"];
    if ($email_count > 0) {
      $text = _("Email address already exists.  Not added.");
      return ($text);
    }

    /* See if the user already exists (better not!) */
    $row = $this->dbManager->getSingleRow("SELECT * FROM users WHERE LOWER(user_name) = LOWER($1) LIMIT 1;",
        array($User), $stmt = __METHOD__ . ".getUserIfExisting");
    if (!empty($row['user_name'])) {
      $text = _("User Exists. Try different Name, User-Name checking is case-insensitive and Duplicate not allowed.");
      return ($text);
    }

    /* check email notification, if empty (box not checked), or if no email
     * specified for the user set to 'n'.
     */
    if(empty($Email_notify)) {
      $Email_notify = '';
    }
    elseif(empty($Email)) {
      $Email_notify = '';
    }

    $ErrMsg = add_user($User,$Desc,$Seed,$Hash,$Perm,$Email,
                       $Email_notify,$agentList,$Folder, $default_bucketpool_fk);

    return ($ErrMsg);
  } // Add()


  public function Output() {
    /* If this is a POST, then process the request. */
    $User = GetParm('username', PARM_TEXT);
    if (!empty($User)) {
      $rc = $this->Add();
      if (empty($rc)) {
        $text = _("User");
        $text1 = _("added");
        $this->vars['message'] = "$text $User $text1.";
      }
      else {
        $this->vars['message'] = $rc;
      }
    }

    $V = "<form name='formy' method='POST'>\n";
    $V.= _("To create a new user, enter the following information:<P />\n");
    $Style = "<tr><td colspan=2 style='background:black;'></td></tr><tr>";
    $V.= "<table style='border:1px solid black; text-align:left; background:lightyellow;' width='75%'>";
    $Val = htmlentities(GetParm('username', PARM_TEXT), ENT_QUOTES);
    $text = _("Username");
    $V.= "$Style<th width='25%' >$text</th>";
    $V.= "<td><input type='text' value='$Val' name='username' size=20></td>\n";
    $V.= "</tr>\n";
    $Val = htmlentities(GetParm('description', PARM_TEXT), ENT_QUOTES);
    $text = _("Description, full name, contact, etc. (optional)");
    $V.= "$Style<th>$text</th>\n";
    $V.= "<td><input type='text' name='description' value='$Val' size=60></td>\n";
    $V.= "</tr>\n";
    $Val = htmlentities(GetParm('email', PARM_TEXT), ENT_QUOTES);
    $text = _("Email address");
    $V .= "$Style<th>$text</th>\n";
    $V.= "<td><input type='text' name='email' value='$Val' size=60></td>\n";
    $V.= "</tr>\n";
    $text = _("Access level");
    $V.= "$Style<th>$text</th>";
    $V.= "<td><select name='permission'>\n";
    $text = _("None (very basic, no database access)");
    $V.= "<option value='" . PLUGIN_DB_NONE . "'>$text</option>\n";
    $text = _("Read-only (read, but no writes or downloads)");
    $V.= "<option selected value='" . PLUGIN_DB_READ . "'>$text</option>\n";
    $text = _("Read-Write (read, download, or edit information)");
    $V.= "<option value='" . PLUGIN_DB_WRITE . "'>$text</option>\n";
    $text = _("Full Administrator (all access including adding and deleting users)");
    $V.= "<option value='" . PLUGIN_DB_ADMIN . "'>$text</option>\n";
    $V.= "</select></td>\n";
    $V.= "</tr>\n";
    $text = _("User root folder");
    $V.= "$Style<th>$text";
    $V.= "</th>";
    $V.= "<td><select name='folder'>";
    $V.= FolderListOption(-1, 0);
    $V.= "</select></td>\n";
    $V.= "</tr>\n";
    $text = _("Password (optional)");
    $V.= "$Style<th>$text</th><td><input type='password' name='pass1' size=20></td>\n";
    $V.= "</tr>\n";
    $text = _("Re-enter password");
    $V.= "$Style<th>$text</th><td><input type='password' name='pass2' size=20></td>\n";
    $V.= "</tr>\n";
    $text = _("E-mail Notification");
    $text1 = _("Check to enable email notification when upload scan completes .");
    $V .= "$Style<th>$text</th><td><input type='checkbox'" .
            "name='enote' value='y' checked='checked'>" .
            "$text1</td>\n";
    $V.= "</tr>\n";
    $text = _("Agents selected by default when uploading");
    $V .= "$Style<th>$text\n</th><td> ";
    $V.= AgentCheckBoxMake(-1, array("agent_unpack", "agent_adj2nest", "wget_agent"));

    $V .= "</td>\n";
    $text = _("Default bucketpool");
    $V.= "$Style<th>$text</th>";
    $V.= "<td>";
    $default_bucketpool_fk = 0;
    $V.= SelectBucketPool($default_bucketpool_fk);
    $V.= "</td>";
    $V .= "</tr>\n";
    $V.= "</table border=0><P />";

    $text = _("Add User");
    $V.= "<input type='submit' value='$text'>\n";
    $V.= "</form>\n";
    return $V;
  }
}
$NewPlugin = new user_add;
